### DISCUSSION TITLE
Allow passing Mojo function as argument to Python function

### DISCUSSION THREAD

**Discussion by mahmoudajawad** [modular/modular#813] [Ideas]
Currently, Mojo doesn't support passing Mojo functions as arguments to Python functions. Certain Python libs accept a `callable` as an argument which runs side effects, and Python interop in Mojo doesn't support this.

Take following example:
```python
# hello.py
from typing import Callable

class MyClass:
  def __init__(self, arg1: str):
    self.arg1 = arg1
  
  def print_arg1(self) -> None:
    print(self.arg1)

def do(func: Callable) -> None:
  func()

if __name__ == "__main__":
  MyClass("Hello, world!").print_arg1()
```

```python
# hello.mojo
from python.python import (
    Python,
    _destroy_python,
    _init_python,
)
from python.object import PythonObject

def print_arg1() -> None:
  Python.add_to_path(".")
  let hello = Python.import_module("hello")
  let my_class = hello.MyClass("Hello, world from Mojo!")
  print(my_class.arg1)

def main():
  Python.add_to_path(".")
  let hello = Python.import_module("hello")
  let my_class = hello.MyClass("Hello, world from Python!")
  print(my_class)
  hello.do(my_class.print_arg1)
  hello.do(print_arg1)
```

Output:
```
hello.mojo:37:11: error: invalid call to '__call__': argument #1 cannot be converted from 'fn() raises -> None' to 'PythonObject'
  hello.do(print_arg1)
  ~~~~~~~~^~~~~~~~~~~~
hello.mojo:1:1: note: function declared here
# ===----------------------------------------------------------------------=== #
^
mojo: error: failed to parse the provided Mojo
```

**Comment by rarebreed** [comment]
It looks like mojo `fn` types are [not automatically convertible to PythonObject types yet](https://docs.modular.com/mojo/programming-manual.html#mojo-types-in-python).

The problem seems to be that specifically in this case, the `print_arg1` _fn_ has a type of `fn() raises -> None`.  What you want to pass to `hello.do()` is a PythonObject type.  

Calling `hello.do(print_arg1)` basically does this:

- The `hello` variable is of type PythonObject.  
- Calling `do` (if it is following python semantics) will call `__getattr__("do")` which also returns a PythonObject.  
- Then the `__call__` method will be called from that PythonObject.
- The type signature for call is  `PythonObject.__call__(inout self, *args: Self)` where Self is PythonObject.  

So the key is that you need to pass in a PythonObject and not the `fn() raises -> None` type that mojo sees `print_arg1` as.

I was looking at the `PythonObject.__init__` constructors, but there isn't one that can take that mojo type.  Surprisingly, there isn't one that can take an `object` type either (the type mojo uses for untyped arguments that are passed to `def` type functions).

**Reply by mahmoudajawad** [reply]
Eventually, in my search I had the same mind-path, and ended-up exactly where you did and tried wrapping different things with `PythonObject` in a hope one would infer the results I am looking for, but not.
